pkg/abstractions/endpoint/buffer.go

```diff
@@ -232,43 +236,62 @@ func (rb *RequestBuffer) discoverContainers() {
 			rb.availableContainers = availableContainers
 			rb.availableContainersLock.Unlock()
 
-			time.Sleep(1 * time.Second)
+			time.Sleep(500 * time.Millisecond)
 		}
 	}
 }
 
-func (rb *RequestBuffer) requestsInFlight(containerId string) (int, error) {
-	val, err := rb.rdb.Get(rb.ctx, Keys.endpointRequestsInFlight(rb.workspace.Name, rb.stubId, containerId)).Int()
+func (rb *RequestBuffer) requestTokens(containerId string) (int, error) {
+	tokenKey := Keys.endpointRequestTokens(rb.workspace.Name, rb.stubId, containerId)
+
+	val, err := rb.rdb.Get(rb.ctx, tokenKey).Int()
 	if err != nil && err != redis.Nil {
 		return 0, err
 	} else if err == redis.Nil {
-		return 0, nil
+		maxTokens := int(rb.stubConfig.Workers)
+
+		err = rb.rdb.SetNX(rb.ctx, tokenKey, maxTokens, 0).Err()
```

Comment:
This function is being called by multiple goroutines and could result in a race condition. Its possible that one or more goroutines could both get redis.Nil and attempt to call `SetNX`, but with just one goroutine succeeding. Even using `Set` here could overwrite another goroutines call to it. 

From what I can tell `SetNX` is atomic, so we should probably call it first with .Result() and check for the error and if the key was created. Then if the key was created, we can return `maxTokens` or if it wasn't we can return `Get().Int()`

Reply:
Yeah probably true. Might even make sense to just create this when the endpoint container is scheduled in instance.go... gotta think about that a little more though

Reply:
Eh, nah I went with the setnx/get. Should be good now.